src/flag_gems/ops/weightnorm.py

```diff
@@ -0,0 +1,150 @@
+    m_idx = col_offset // v_shape2
+    n_idx = row_offset
+    k_idx = col_offset % v_shape2
+
+    mask = m_idx < v_shape0 and row_mask
+
+    v_offsets = m_idx * v_stride0 + n_idx * v_stride1 + k_idx * v_stride2
+    v_value = tl.load(v + v_offsets, mask = mask)
+    v_block += v_value * v_value
+  v_sum = tl.sum(v_block, axis = 1) + eps
+
+
+  for base in range(0, o_shape0 * o_shape2, BLOCK_COL_SIZE):
+    col_offset = base + tid_n
+    m_idx = col_offset // o_shape2
+    n_idx = row_offset
+    k_idx = col_offset % o_shape2
+
+    mask = m_idx < o_shape0 and row_mask
+
+    v_offsets = (m_idx % vb_shape0) * vb_stride0 + (n_idx % vb_shape1) * vb_stride1 + (k_idx % vb_shape2) * vb_stride2
+    v_value = tl.load(v_broad + v_offsets, mask = mask)
+    v_vec = rsqrt(v_sum[:, None]) * v_value
+
+    g_offset = (m_idx % gb_shape0) * gb_stride0 + (n_idx % gb_shape1) * gb_stride1 + (k_idx % gb_shape2) * gb_stride2
+    g_value = tl.load(g_broad + g_offset, mask = mask)
+    out = v_vec * g_value
+    out_offset = m_idx * o_shape1 * o_shape2 +  n_idx * o_shape2 + k_idx
+    tl.store(output + out_offset, out, mask = mask)
+
+def weight_norm(v, g, dim = 0):
+  logging.debug("GEMS WEIGHTNORM")
+
+  v = v.contiguous()
+  g = g.contiguous()
+  dim_neg = dim - len(v.shape)
+  output_shape = torch.broadcast_shapes(v.shape, g.shape)
+  output = torch.empty(output_shape, device = v.device, dtype = v.dtype)
+  
+  v_broad = v
+  g_broad = g
+  v_re_shape = (v.shape[:dim_neg] + v.shape[dim_neg+1:])
+  g_re_shape = (g.shape[:dim_neg] + g.shape[dim_neg+1:])
+  for i in range(len(v_re_shape)-1, 0, -1):
+    if v_re_shape[i] == 1 and sum(v_re_shape[:i]) != i:
+      v_broad = torch.broadcast_to(v, output_shape).clone()
+      break
+  for i in range(len(g_re_shape)-1, 0, -1):
+    if g_re_shape[i] == 1 and sum(g_re_shape[:i]) != i:
+      g_broad = torch.broadcast_to(g, output_shape).clone()
+      break
+
+  o_g_v_vb_gb = [[math.prod(x.shape[:dim_neg]), x.shape[dim_neg], math.prod(x.shape[dim_neg+1:])] for x in [output, g, v, v_broad, g_broad]]
+
+  step = [1] * 3
+  v_vb_gb = [[0] * 3, [0] * 3, [0] * 3]
+  for i in range(2, -1, -1):
+    for j in range(len(v_vb_gb)):
+      if o_g_v_vb_gb[j + 2][i] != 1:
+        v_vb_gb[j][i] = step[j]
+      step[j] = step[j] * o_g_v_vb_gb[j + 2][i]
+
+  grid = lambda META: (triton.cdiv(o_g_v_vb_gb[0][1], META["BLOCK_ROW_SIZE"]),)
+  with torch.cuda.device(v.device):
+    weight_norm_kernel[grid](
+      output,
+      v,
+      v_broad,
+      g_broad,
+      o_g_v_vb_gb[0][0], o_g_v_vb_gb[0][1], o_g_v_vb_gb[0][2],
+      o_g_v_vb_gb[2][0], o_g_v_vb_gb[2][2],
+      o_g_v_vb_gb[3][0], o_g_v_vb_gb[3][1], o_g_v_vb_gb[3][2],
+      o_g_v_vb_gb[4][0], o_g_v_vb_gb[4][1], o_g_v_vb_gb[4][2],
+      v_vb_gb[0][0], v_vb_gb[0][1], v_vb_gb[0][2],
+      v_vb_gb[1][0], v_vb_gb[1][1], v_vb_gb[1][2],
+      v_vb_gb[2][0], v_vb_gb[2][1], v_vb_gb[2][2],
+      eps = torch.finfo(torch.float32).tiny
+    )
+  return output
```

Comment:
The aten native function weight_norm_cuda is what's to be aligned with so we should return w, norms tuple.